services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/QueueAttributesManager.java

```diff
@@ -0,0 +1,157 @@
+        this.queueUrl = queueUrl;
+    }
+
+    /**
+     * Retrieves the received message timeout based on the provided request and queue attributes.
+     *
+     * @param rq                 The receive message request
+     * @param configuredWaitTime The configured minimum wait time
+     * @return CompletableFuture with the calculated receive message timeout in milliseconds
+     */
+    public CompletableFuture<Duration> getReceiveMessageTimeout(ReceiveMessageRequest rq, Duration configuredWaitTime) {
+        Integer waitTimeSeconds = rq.waitTimeSeconds();
+        if (waitTimeSeconds != null) {
+            long waitTimeMillis = TimeUnit.SECONDS.toMillis(waitTimeSeconds);
+            return CompletableFuture.completedFuture(Duration.ofMillis(Math.max(configuredWaitTime.toMillis(), waitTimeMillis)));
+        }
+
+        CompletableFuture<Map<QueueAttributeName, String>> attributeFuture = getAttributeMap();
+        CompletableFuture<Duration> resultFuture = attributeFuture.thenApply(attributes -> {
+            String waitTimeSecondsStr = attributes.get(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS);
+            long waitTimeFromSqsMillis = TimeUnit.SECONDS.toMillis(Long.parseLong(waitTimeSecondsStr));
+            return Duration.ofMillis(Math.max(configuredWaitTime.toMillis(), waitTimeFromSqsMillis));
+        });
+
+        resultFuture.whenComplete((r, t) -> {
+            if (resultFuture.isCancelled()) {
+                attributeFuture.cancel(true);
+            }
+        });
+
+        return resultFuture;
+    }
+
+    /**
+     * Retrieves the visibility timeout for the queue.
+     *
+     * @return CompletableFuture with the visibility timeout in nanoseconds
+     */
+    public CompletableFuture<Duration> getVisibilityTimeout() {
+        CompletableFuture<Map<QueueAttributeName, String>> attributeFuture = getAttributeMap();
+        CompletableFuture<Duration> resultFuture = attributeFuture.thenApply(attributes -> {
+            String visibilityTimeoutStr = attributes.get(QueueAttributeName.VISIBILITY_TIMEOUT);
+            return Duration.ofSeconds(Integer.parseInt(visibilityTimeoutStr));
+        });
+
+        resultFuture.whenComplete((r, t) -> {
+            if (resultFuture.isCancelled()) {
+                attributeFuture.cancel(true);
+            }
+        });
+
+        return resultFuture;
+
+    }
+
+    /**
+     * Retrieves the queue attributes based on the predefined attribute names.
+     *
+     * @return CompletableFuture with the map of attribute names and their values.
+     */
+    private CompletableFuture<Map<QueueAttributeName, String>> getAttributeMap() {
+        CompletableFuture<Map<QueueAttributeName, String>> future = queueAttributeMap.get();
+
+        if (future == null || future.isCompletedExceptionally()) {
+            CompletableFuture<Map<QueueAttributeName, String>> newFuture = fetchQueueAttributes();
+            if (queueAttributeMap.compareAndSet(future, newFuture)) {
```

Comment:
Should we cancel `newFuture` if this comparison fails?

Reply:
yeap good catch , updated the implementation to reduce redundant call if another call is in progress